pkg/node/chain.go

```diff
@@ -257,3 +260,77 @@ func InitSwap(
 
 	return swapService, priceOracle, nil
 }
+
+func GetTxHash(stateStore storage.StateStorer, logger logging.Logger, trxString string) ([]byte, error) {
+
+	if trxString != "" {
+		txHashTrimmed := strings.TrimPrefix(trxString, "0x")
+		if len(txHashTrimmed) != 64 {
+			return nil, errors.New("invalid length")
+		}
+		txHash, err := hex.DecodeString(txHashTrimmed)
+		if err != nil {
+			return nil, err
+		}
+		logger.Infof("using the provided transaction hash %x", txHash)
+		return txHash, nil
+	}
+
+	var txHash common.Hash
+	key := chequebook.ChequebookDeploymentKey
+	if err := stateStore.Get(key, &txHash); err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			return nil, errors.New("chequebook deployment transaction hash not found, please specify the transaction hash manually")
+		}
+		return nil, err
+	}
+
+	logger.Infof("using the chequebook transaction hash %x", txHash)
+	return txHash.Bytes(), nil
+}
+
+func GetTrxNextBlock(ctx context.Context, logger logging.Logger, stateStore storage.StateStorer, backend transaction.Backend, monitor transaction.Monitor, trx []byte, blockHash string) ([]byte, error) {
+
+	if blockHash != "" {
+		blockHashTrimmed := strings.TrimPrefix(blockHash, "0x")
+		if len(blockHashTrimmed) != 64 {
+			return nil, errors.New("invalid length")
+		}
+		blockHash, err := hex.DecodeString(blockHashTrimmed)
+		if err != nil {
+			return nil, err
+		}
+		logger.Infof("using the provided block hash %x", blockHash)
+		return blockHash, nil
+	}
+
+	// check statestore first
+	var hash common.Hash
+	if err := stateStore.Get(nextBlockHashKey, &hash); err != nil {
+		if !errors.Is(err, storage.ErrNotFound) {
+			return nil, err
+		}
+	} else {
+		return hash.Bytes(), nil
+	}
+
+	// if not found in statestore, fetch from chain
+	tx, err := backend.TransactionReceipt(ctx, common.BytesToHash(trx))
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := monitor.WaitBlock(ctx, big.NewInt(0).Add(tx.BlockNumber, big.NewInt(1)))
+	if err != nil {
+		return nil, err
+	}
+
+	hash = block.Hash()
+
+	err = stateStore.Put(nextBlockHashKey, hash)
```

Comment:
We are storing both the transaction and blockhash in the state. Is one of these perhaps redundant?

Reply:
> We are storing both the transaction and blockhash in the state. Is one of these perhaps redundant?

there can be several txs by the same account in the parent block while there is only a single child block, so the blockhash is the redundant one

Reply:
It stores it so that the next boot up does not need to call the blockchain to fetch it.

Reply:
I think it's good as it is.